### PR TITLE
Added args parameter to LightningCLI to ease running from within Python

### PR DESCRIPTION
**Pull request by mauvilsa**
## What does this PR do?

As the title says and more details in the updated docs. This was recently discussed in the lightning-cli slack channel and has come up a few times in comments and discussions.

Fixes #13277

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source-pytorch/cli/lightning_cli_advanced_3.rst [hunk 1]
```diff
@@ -354,3 +354,33 @@ You can also pass the class path directly, for example, if the optimizer hasn't
...
+Run from Python
+^^^^^^^^^^^^^^^
+
+Even though the :class:`~pytorch_lightning.cli.LightningCLI` class is designed to help in the implementation of command
+line tools, for some use cases it is desired to run directly from Python. To support these use cases, the ``args``
+parameter can be used, for example:
+
+.. testcode::
+
+    cli = LightningCLI(MyModel, args=["--trainer.max_epochs=100", "--model.encoder_layers=24"])
```
Comment (carmocca) [inline docs/source-pytorch/cli/lightning_cli_advanced_3.rst:368]: To clarify, are these merged with the normal CLI args? And `--config=...`? If so, which takes precedence?

File: docs/source-pytorch/cli/lightning_cli_advanced_3.rst [hunk 2]
```diff
@@ -354,3 +354,33 @@ You can also pass the class path directly, for example, if the optimizer hasn't
...
+also possible to provide to ``args`` a ``dict`` or `Namespace
+<https://jsonargparse.readthedocs.io/en/stable/#jsonargparse.Namespace>`__. For example:
+
+.. testcode::
+
+    cli = LightningCLI(
+        MyModel,
+        args={
+            "trainer": {
+                "max_epochs": 100,
```
Comment (carmocca) [inline docs/source-pytorch/cli/lightning_cli_advanced_3.rst:380]: Does this make `trainer_defaults` redundant? What if both are set?

File: docs/source-pytorch/cli/lightning_cli_advanced_3.rst (hunk 1 above)
Comment (mauvilsa) [reply docs/source-pytorch/cli/lightning_cli_advanced_3.rst:368]: They are not merged. If `args != None` and this is used from the command line, all the command line arguments given would be ignored. `args` is only to be able to run things from Python like if it were from the command line.

File: docs/source-pytorch/cli/lightning_cli_advanced_3.rst (hunk 2 above)
Comment (mauvilsa) [reply docs/source-pytorch/cli/lightning_cli_advanced_3.rst:380]: Not redundant. `args` as a dict is like running from command line `cli.py --config={...}` and it can include all options, `subcommand`, `data`, `model`, `trainer`. I do see that this needs to be explained better in the docs to understand the difference. I will improve it.

File: docs/source-pytorch/cli/lightning_cli_advanced_3.rst (hunk 1 above)
Comment (carmocca) [reply docs/source-pytorch/cli/lightning_cli_advanced_3.rst:368]: In that case, can we check if command line args were also passed and then raise an error?

I imagine people will expect it to work and would be confused if they are ignored.

Comment (mauvilsa) [reply docs/source-pytorch/cli/lightning_cli_advanced_3.rst:368]: Good idea!

File: docs/source-pytorch/cli/lightning_cli_advanced_3.rst [hunk 3]
```diff
@@ -354,3 +354,33 @@ You can also pass the class path directly, for example, if the optimizer hasn't
...
+line tools, for some use cases it is desired to run directly from Python. To support these use cases, the ``args``
+parameter can be used, for example:
+
+.. testcode::
+
+    cli = LightningCLI(MyModel, args=["--trainer.max_epochs=100", "--model.encoder_layers=24"])
+
+All the features that are supported from the command line can be used when giving ``args`` as a list of strings. It is
+also possible to provide to ``args`` a ``dict`` or `Namespace
+<https://jsonargparse.readthedocs.io/en/stable/#jsonargparse.Namespace>`__. For example:
```
Comment (awaelchli) [inline docs/source-pytorch/cli/lightning_cli_advanced_3.rst:372]: Will a regular argparse Namespace also work?

**Review by awaelchli — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: docs/source-pytorch/cli/lightning_cli_advanced_3.rst (hunk 3 above)
Comment (mauvilsa) [reply docs/source-pytorch/cli/lightning_cli_advanced_3.rst:372]: No, a regular argparse Namespace does not work. I will change the text to say `jsonargparse.Namespace`.

File: docs/source-pytorch/cli/lightning_cli_advanced_3.rst [hunk 4]
```diff
@@ -360,27 +360,49 @@ Run from Python
...
+        "trainer": {
+            "max_epochs": 100,
         },
-    )
+        "model": {},
+    }
+
+    for encoder_layers in [8, 16, 24]:
+        args["model"]["encoder_layers"] = encoder_layers
+        cli_main(args)
```
Comment (carmocca) [inline docs/source-pytorch/cli/lightning_cli_advanced_3.rst:404]: I don't think this example makes a lot of sense. When would you instantiate the CLI like this in a for-loop?

File: src/pytorch_lightning/cli.py [hunk 5]
```diff
@@ -478,10 +482,14 @@ def link_optimizers_and_lr_schedulers(parser: LightningArgumentParser) -> None:
                 add_class_path = _add_class_path_generator(class_type)
                 parser.link_arguments(key, link_to, compute_fn=add_class_path)
 
-    def parse_arguments(
-        self, parser: LightningArgumentParser, args: Union[List[str], Dict[str, Any], Namespace]
-    ) -> None:
+    def parse_arguments(self, parser: LightningArgumentParser, args: ArgsType) -> None:
         """Parses command line arguments and stores it in ``self.config``."""
+        if args is not None and len(sys.argv) > 1:
+            raise MisconfigurationException(
```
Comment (carmocca) [inline src/pytorch_lightning/cli.py:488]: We are moving away from this custom exception class

```suggestion
            raise ValueError(
```

File: tests/tests_pytorch/test_cli.py [hunk 6]
```diff
@@ -1414,7 +1414,15 @@ def test_pytorch_profiler_init_args():
...
+    with mock.patch("sys.argv", [""]):
+        cli = LightningCLI(TestModel, run=False, args=args)
     assert isinstance(cli.model, TestModel)
     assert cli.config.trainer.logger is False
     assert cli.model.foo == 456
+
+
+def test_lightning_cli_args_and_sys_argv_exception():
+    with mock.patch("sys.argv", ["", "--model.foo=456"]), pytest.raises(
+        MisconfigurationException, match="LightningCLI's args parameter "
```
Comment (carmocca) [inline tests/tests_pytorch/test_cli.py:1428]:
```suggestion
        ValueError, match="LightningCLI's args parameter "
```

**Review by carmocca — APPROVED** [review]

File: docs/source-pytorch/cli/lightning_cli_intermediate.rst [hunk 7]
```diff
@@ -87,8 +87,12 @@ The simplest way to control a model with the CLI is to wrap it in the LightningC
     # simple demo classes for your convenience
     from pytorch_lightning.demos.boring_classes import DemoModel, BoringDataModule
 
-    cli = LightningCLI(DemoModel, BoringDataModule)
-    # note: don't call fit!!
+    def cli_main():
+        cli = LightningCLI(DemoModel, BoringDataModule)
+        # note: don't call fit!!
+
+    if __name__ == "__main__":
```
[the remark this answers is not in the document]
Comment (mauvilsa) [reply docs/source-pytorch/cli/lightning_cli_intermediate.rst:96]: It is not a requirement. Only a good practice.

File: docs/source-pytorch/cli/lightning_cli_advanced_3.rst (hunk 4 above)
Comment (mauvilsa) [reply docs/source-pytorch/cli/lightning_cli_advanced_3.rst:404]: In a jupyter notebook someone might want to run the cli varying the number of encoding layers. It could be some other example without a loop. The point is to show that giving a dict makes more sense in some cases.

File: docs/source-pytorch/cli/lightning_cli_intermediate.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (mauvilsa) [reply docs/source-pytorch/cli/lightning_cli_intermediate.rst:96]: We should add a comment saying that this is a good practice.

[the remark this answers is not in the document]
Comment (mauvilsa) [reply docs/source-pytorch/cli/lightning_cli_intermediate.rst:96]: Done

File: docs/source-pytorch/cli/lightning_cli_advanced_3.rst (hunk 4 above)
Comment (mauvilsa) [reply docs/source-pytorch/cli/lightning_cli_advanced_3.rst]: Changed the example
